cody: initial vs-code-0-14 (#6517)
sourcegraph/about: content/blogposts/2023/cody-vscode-0-14-release.md

---
title: "Cody for VS Code v0.12 release"
authors:
  - name: Kalan Chan
    url: https://handbook.sourcegraph.com/team/#kalan-chan
  - name: Tim Lucas
    url: https://handbook.sourcegraph.com/team/#tim-lucas
publishDate: 2023-10-04T10:00-07:00

> description: "Smart selection for `/doc` and `/test`, Inline Documentation with /doc, New custom command modes, New code block actions in chat"

Author: Just plain sentence case here (matching the previous posts) with a full stop. I've updated to match the suggested heading edits below.

```suggestion
description: "Smart selection for /doc and /test, improvements to /doc, a new mode configuration for custom commands, and updated code block action buttons in chat."
```

tags: [blog]
slug: "cody-vscode-0-14-release"
published: true
heroImage: https://storage.googleapis.com/sourcegraph-assets/blog/cody-vscode-0-14-release/cody-vscode-0.14.0-og-image.jpg
socialImage: https://storage.googleapis.com/sourcegraph-assets/blog/cody-vscode-0-14-release/cody-vscode-0.14.0-og-image.jpg
---

[Cody for VS Code](https://marketplace.visualstudio.com/items?itemName=sourcegraph.cody-ai) v0.14 is now available and includes improvements to the `/doc` and `/test` commands, a new mode configuration for custom commands, and experimental smart selection.

> ## Smart selection for `/doc` and `/test`

Author: Let's leave off the code tags in headings because it renders a bit weird:

```suggestion
## Smart selection for /doc and /test
```

> You can now use the `/doc` and `/test` commands without selecting any code first, and Cody will use the position of your cursor to guess out what you're wanting to document or code.

Author: ```suggestion
You can now use the `/doc` and `/test` commands without selecting any code first, and Cody will use the position of your cursor to guess out what you're wanting to document or code.

For unit tests, it means you can now generate them by right clicking on a function name and selecting "Cody → Generate Unit Tests". Or using the keyboard, anywhere in a function, by pressing option-c (Mac) or alt-c (Windows/Linux) and running `/test`.

<img alt="Screenshot of an smart selection" src="https://storage.googleapis.com/sourcegraph-assets/blog/cody-vscode-0-14-release/smart-select.png" style={{marginTop:"1rem",marginBottom:"1.5rem",marginLeft:0,maxWidth:628}} />

This feature is still under development, please give it a test and provide feedback.

> ## Inline Documentation with /doc
> 
> The `/doc` command now adds the generated documentation directly above your selected code. You no longer need to copy and paste the generated documentation from the chat view.
> 
> <img alt="Screenshot of an inline doc generation" src="https://storage.googleapis.com/sourcegraph-assets/blog/cody-vscode-0-14-release/inline-doc-2.png" style={{marginTop:"1rem",marginBottom:"1.5rem",marginLeft:0,maxWidth:628}} />
> 

Author: We're missing the new doc prompt changes here that just landed too.

```suggestion
## Improvements to /doc

The `/doc` command now adds the generated documentation directly above your selected code. You no longer need to copy and paste the generated documentation from the chat view.

 <img alt="Screenshot of an inline doc generation" src="https://storage.googleapis.com/sourcegraph-assets/blog/cody-vscode-0-14-release/inline-doc-2.png" style={{marginTop:"1rem",marginBottom:"1.5rem",marginLeft:0,maxWidth:628}} />
 
We've also improved the `/doc` command’s underlying prompt so it now better matches your existing comment styles, generates more concise explanations, and better understands when to use a docstring-style API comment vs a plain code comment.
```

## New custom command modes 

Custom commands now support a new `mode` configuration field that allows commands to be customized for how Cody should execute them:

- `inline` (default) - Runs the prompt in the inline chat window 
- `edit` - Runs the prompt on the selected code for refactoring workflows
- `insert` - Runs the prompt and inserts the response on top of the selection

> This enables custom commands to be tailored for specific use cases like refactoring or documentation generation. See [docs](https://docs.sourcegraph.com/cody/custom-commands) for more details.

Author: ```suggestion
See the [custom commands documentation](https://docs.sourcegraph.com/cody/custom-commands) for details on the new `mode` property.
```

> ## New code block actions in chat
> 
> We added a new "Save to File..." button on code blocks in the chat view. This allows you to easily extract and save code Cody generates to a file in your project.

Author: It's probably good to make it really obvious that we've moved the buttons, so perhaps we'll update the header and copy to focus on that first, and "Save to File" second.

```suggestion
## New Copy, Insert and Save to File buttons

The "Copy" and "Insert at Cursor" actions on chat code blocks are now always visible, accessible via a new bar on the bottom of the code blocks.

We've also added a new "Save to File..." button, allowing you to save the code block directly to a new file in your project.
```

  - name: Tim Lucas
    url: https://handbook.sourcegraph.com/team/#tim-lucas
  - name: Beatrix Woo
    url: https://handbook.sourcegraph.com/team/#beatrix-woo  
publishDate: 2023-10-04T10:00-07:00
description: "inline /doc generation, new custom command modes, smart selection context, and saving chat code to files."
tags: [blog]
slug: "cody-vscode-0-14-release"
published: true
heroImage: https://storage.googleapis.com/sourcegraph-assets/cody-vscode-0.14.0-og-image.jpg
socialImage: https://storage.googleapis.com/sourcegraph-assets/cody-vscode-0.14.0-og-image.jpg
---

[Cody for VS Code](https://marketplace.visualstudio.com/items?itemName=sourcegraph.cody-ai) v0.14 is now available and includes improvements to the `/doc` and `/test` commands, a new mode configuration for custom commands, and experimental smart selection.

## Inline Documentation with /doc

Previously when using the `/doc` command, the generated documentation would be shown in the chat view. It will now directly insert the documentation text right above the selected code in your editor. This makes it easier to quickly add documentation stubs as you code.

<!-- TODO add example screenshot here -->

> ## Experimental smart selection
> 
> We've started work on a smart selection capability that will automatically determine what code to run the `/doc` and `/test` commands on based on heuristics like scope and recent activity. This will remove the need to manually highlight code before generating documentation or tests.
> 
> Smart selection is still experimental but we're excited about the potential to streamline workflows. 

Author: It's okay to cover behind the scenes, but "We've started work" sounds like maybe there's no changes in this release and it's an FYI. And experimental things are normally opt-in and buggy, and we haven't pointed out how you see/use this new feature, or how you avoid this new feature if you don't want the bugs.

Are there caveats to this feature we should point out? Language support? Known bugs? Do we want feedback if people find it's not working well for them?

What about something more like…

```suggestion
## Smart selection for `/doc` and `/test`

You can now use the `/doc` and `/test` commands without selecting any code first, and Cody will use the position of your cursor to guess out what you're wanting to document or code.

(image)

This feature is still under development, and doesn't yet handle cases xyz. Please give it a test and provide feedback.
```

<!-- TODO add example screenshot here -->

## New custom command modes 

Custom commands now support a new `mode` configuration field that allows commands to be customized for how Cody should execute them:

- `inline` (default) - Runs the prompt in the inline chat window 
- `edit` - Runs the prompt on the selected code for refactoring workflows

> - `insert` - Runs the prompt and inserts the response on top of the selection

Reviewer: @abeatrix will update this after

> - `insert` - Runs the prompt and inserts the response on top of the selection

Author: Opened one here: https://github.com/sourcegraph/sourcegraph/pull/57328

We need to update the words in the blog post to match

> - `insert` - Runs the prompt and inserts the response on top of the selection

Author: We still need to update these words about the mode options, and the defaults, to match the docs @kalanchan (inline isn’t the default)
